### PR TITLE
[BACKLOG-41451] Implement folder move, fix PathStyle and path parsing

### PR DESCRIPTION
**Pull request by tgf**

### CODE REVIEW

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java [hunk 1]
```diff
@@ -411,6 +413,23 @@ protected void doRename( FileObject newFile ) throws Exception {
     delete();
   }
 
+  private void doFolderMove( FileObject sourceFolder, FileObject targetFolder ) throws FileSystemException {
```
Comment (abryant-hv) [inline s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:416]: Would love to have automation actually testing this. I wonder how we could do that. Are there any steps or job entries that move folders?

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java [hunk 2]
```diff
@@ -411,6 +413,23 @@ protected void doRename( FileObject newFile ) throws Exception {
...
+    FileObject[] children = sourceFolder.getChildren();
+    for ( FileObject child : children ) {
+      FileObject targetChild = targetFolder.resolveFile( child.getName().getBaseName() );
+      if ( child.isFolder() ) {
+        // logger.trace( "moving folder recursive [{}] -> [{}]", child.getPath(), targetChild.getPath() );
+        doFolderMove( child, targetChild );
+      } else if ( child.isFile() ) {
+        logger.trace( "moving file [{}] -> [{}]", child.getPublicURIString(), targetChild.getPublicURIString() );
+        child.moveTo( targetChild );
+      }
```
Comment (abryant-hv) [inline s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:428]: does the order here matter? Like, do we need to create the target folders before moving the child files?

File: s3-vfs/src/test/java/org/pentaho/s3/vfs/S3MoveTest.java [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    assertEquals( endpoint, cfg.getEndpoint() );
+    assertEquals( pathStyleAccess, cfg.getPathStyleAccess() );
+    assertEquals( sigVersion, cfg.getSignatureVersion() );
+  }
+
+  @Test
+  public void testMoveFile() throws Exception {
+    final String bucket = "balde";
+    S3FileName root = new S3FileName( "s3", "balde", "/balde/dir", FileType.FOLDER );
+    TestAmazonS3 s3 = new TestAmazonS3();
```
Comment (abryant-hv) [inline s3-vfs/src/test/java/org/pentaho/s3/vfs/S3MoveTest.java:106]: this is awesome

File: s3-vfs/src/test/java/org/pentaho/s3n/vfs/S3NFileObjectTest.java [hunk 4]
```diff
@@ -145,13 +144,13 @@ public void setUp() throws Exception {
     when( s3ObjectMetadata.getLastModified() ).thenReturn( testDate );
     when( s3ServiceMock.getObject( anyString(), anyString() ) ).thenReturn( s3ObjectMock );
 
-    when( fileSystemSpy.getS3Client() ).thenReturn( s3ServiceMock );
+    // when( fileSystemSpy.getS3Client() ).thenReturn( s3ServiceMock );
```
Comment (abryant-hv) [inline s3-vfs/src/test/java/org/pentaho/s3n/vfs/S3NFileObjectTest.java:147]: remove?

**Review by abryant-hv — APPROVED** [review]

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java (hunk 2 above)
Comment (tgf) [reply s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:428]: No, I think it's only really necessary for empty folders

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java (hunk 1 above)
Comment (tgf) [reply s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:416]: There is a move entry that could work with vfs. And we can always code some integration tests

**Comment by tgf** [comment]
@abryant-hv added S3CommonFileObject.getPath since last review

**Comment by abryant-hv** [comment]
> @abryant-hv added S3CommonFileObject.getPath since last review

Seems fine. :)

**Review by bconlon-hv — APPROVED** [review]

**Pull request merged**